File: opensource/kernel/src/main/java/com/sos/scheduler/engine/kernel/log/PrefixLog.java
package com.sos.scheduler.engine.kernel.log;

import com.sos.scheduler.engine.cplusplus.runtime.Sister;
import com.sos.scheduler.engine.cplusplus.runtime.SisterType;
import com.sos.scheduler.engine.cplusplus.runtime.annotation.ForCpp;
import com.sos.scheduler.engine.kernel.cppproxy.Prefix_logC;
import org.apache.log4j.Logger;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

@ForCpp
public final class PrefixLog implements Sister, SchedulerLogger {
    private static final Logger logger = Logger.getLogger(PrefixLog.class);

    private final Prefix_logC cppProxy;
    private final List<LogSubscription> subscriptions = new ArrayList<LogSubscription>();
    private final LogSubscription[] emptyLogSubscriptions = new LogSubscription[0];
    private LogSubscription[] subscriptionsSnapshot = emptyLogSubscriptions;
    private boolean subscriptionsModified = false;
//    private File fileToBeRemoved = null;

    public PrefixLog(Prefix_logC cppProxy) {
        this.cppProxy = cppProxy;
    }

    @Override public void onCppProxyInvalidated() {}

    @ForCpp public void onStarted() {
        for (LogSubscription o: subscriptionsSnapshot()) {
            try {
                o.onStarted();
            } catch (Throwable t) { logger.error(o+": "+t, t); }
        }
    }

    @ForCpp public void onClosed() {
        for (LogSubscription o: subscriptionsSnapshot()) {
            try {
                o.onClosed();
            } catch (Throwable t) { logger.error(o+": "+t, t); }
        }
        subscriptions.clear();
    }

    @ForCpp public void onLogged() {
        for (LogSubscription o: subscriptionsSnapshot()) {
            try {
                o.onLogged();
            } catch (Throwable t) { logger.error(o+": "+t, t); }
        }
    }

//    @ForCpp public void removeFile(String file) {
//        fileToBeRemoved = new File(file);
//    }

//    private void removeFileNow() {
//        if (fileToBeRemoved != null) {
//            boolean ok = fileToBeRemoved.delete();
//            if (!ok)
//                logger.error("SCHEDULER-291 "+fileToBeRemoved);
//        }
//    }

    private LogSubscription[] subscriptionsSnapshot() {
        synchronized (subscriptions) {
            if (subscriptionsModified) {
                subscriptionsModified = false;
                subscriptionsSnapshot = subscriptions.isEmpty()?
                        emptyLogSubscriptions : subscriptions.toArray(new LogSubscription[subscriptions.size()]);
            }
        }
        return subscriptionsSnapshot;
    }

    public void subscribe(LogSubscription o) {
        synchronized (subscriptions) {
            subscriptions.add(o);
            subscriptionsModified = true;
        }
    }

    public void unsubscribe(LogSubscription o) {
        synchronized (subscriptions) {
            subscriptions.remove(o);
            subscriptionsModified = true;
//            if (subscriptions.isEmpty())
//                removeFileNow();
        }
    }

    @Override public void info(String s) {
        cppProxy.info(s);
    }

    @Override public void warn(String s) {
        cppProxy.warn(s);
    }

    @Override public void error(String s) {
        cppProxy.error(s);
    }

    public void debug3(String s) {
        cppProxy.debug3(s);
    }

    @Override public void debug(String s) {
        debug3(s);
    }

    /** @return "", wenn für den Level keine Meldung vorliegt. */
    public String lastByLevel(SchedulerLogLevel level) {
        return cppProxy.java_last(level.getCppName());
    }

    public boolean isStarted() {
        return cppProxy.started();
    }

    public File getFile() {
        return new File(cppProxy.this_filename());
    }

    public static class Type implements SisterType<PrefixLog, Prefix_logC> {
        @Override public final PrefixLog sister(Prefix_logC proxy, Sister context) {
            return new PrefixLog(proxy);
        }
    }
}

File: engine-kernel/src/main/scala/com/sos/scheduler/engine/kernel/log/PrefixLog.scala
package com.sos.scheduler.engine.kernel.log

import com.sos.jobscheduler.common.guice.GuiceImplicits.RichInjector
import com.sos.jobscheduler.common.scalautil.{Logger, ScalaConcurrentHashSet, SetOnce}
import com.sos.jobscheduler.data.log.{SchedulerLogLevel, SchedulerLogger}
import com.sos.scheduler.engine.cplusplus.runtime.annotation.ForCpp
import com.sos.scheduler.engine.cplusplus.runtime.{Sister, SisterType}
import com.sos.scheduler.engine.kernel.Scheduler
import com.sos.scheduler.engine.kernel.async.SchedulerThreadCallQueue
import com.sos.scheduler.engine.kernel.async.SchedulerThreadFutures.inSchedulerThread
import com.sos.scheduler.engine.kernel.cppproxy.Prefix_logC
import com.sos.scheduler.engine.kernel.scheduler.HasInjector
import java.io.File
import scala.util.control.NonFatal

/**
 * @author Joacim Zschimmer
 */
@ForCpp
final class PrefixLog(
  cppProxy: Prefix_logC,
  schedulerThreadCallQueueOption: Option[SchedulerThreadCallQueue])
extends Sister with SchedulerLogger {

  private val schedulerThreadCallQueueOnce = new SetOnce[SchedulerThreadCallQueue]

  schedulerThreadCallQueueOption foreach { schedulerThreadCallQueueOnce := _ }

  private implicit def schedulerThreadCallQueue = schedulerThreadCallQueueOnce getOrElse {
    throw new IllegalStateException("PrefixLog has not get a SchedulerThreadCallQueue") }

  private val subscriptions = new ScalaConcurrentHashSet[LogSubscription]()

  def onCppProxyInvalidated() = {}

  @ForCpp private def onStarted(scheduler: Scheduler): Unit = {
    if (schedulerThreadCallQueueOnce.isEmpty) {
      schedulerThreadCallQueueOnce.trySet(scheduler.injector.instance[SchedulerThreadCallQueue])
    }
    forAllSubscriptions { _.onStarted() }
  }

  @ForCpp private def onClosed(): Unit = {
    forAllSubscriptions { _.onClosed() }
    subscriptions.clear()
  }

  @ForCpp private def onLogged(): Unit = forAllSubscriptions { _.onLogged() }

  private def forAllSubscriptions(f: LogSubscription ⇒ Unit): Unit =
    for (o ← subscriptions) {
      try f(o)
      catch {
        case NonFatal(t) ⇒ PrefixLog.logger.error(s"$o: $t", t)
      }
    }

  def subscribe(o: LogSubscription): Unit = subscriptions += o

  def unsubscribe(o: LogSubscription): Unit = subscriptions -= o

  def log(level: SchedulerLogLevel, s: String): Unit = inSchedulerThread { cppProxy.java_log(level.cppNumber, s) }

  /** @return "", wenn für den Level keine Meldung vorliegt. */
  def lastByLevel(level: SchedulerLogLevel): String = inSchedulerThread { cppProxy.java_last(level.cppName) }

  def isStarted: Boolean = inSchedulerThread { cppProxy.started }

  def file: File = inSchedulerThread { new File(cppProxy.this_filename) }
}

object PrefixLog {
  private val logger = Logger(getClass)

  class Type extends SisterType[PrefixLog, Prefix_logC] {
    final def sister(proxy: Prefix_logC, context: Sister): PrefixLog = {
      val schedulerThreadCallQueueOption = Option(context) map { _.asInstanceOf[HasInjector].injector.instance[SchedulerThreadCallQueue] }
      new PrefixLog(proxy, schedulerThreadCallQueueOption)
    }
  }
}
